[contain-intrinsic-size] failures related to replaced elements

cathiechen opened the pull request:
#### 43b73e296e5003e669fcb0ba76304837e90cbaa9
<pre>
[contain-intrinsic-size] failures related to replaced elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=253013">https://bugs.webkit.org/show_bug.cgi?id=253013</a>

Reviewed by Alan Baradlay.

This patch makes intrinsicSize() considering both contain: size and inline-size,
so that there is no need to handle contain: inline-size specially.
In computeReplacedLogicalWidth/Height, hasIntrinsicWidth/Height is determined by width/height &gt; 0,
for contain: size/inline-size, hasIntrinsicWidth/Height should be true regardless of the value.
For RenderVideo, in calculateIntrinsicSize(), the intrinsic size should consider contain: inline-size,
so this patch adjustSizeForInlineSizeContainment before return.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-028-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-logical-003-expected.txt:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::intrinsicLogicalWidth const):
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::computeAspectRatioInformationForRenderBox const):
(WebCore::RenderReplaced::computeIntrinsicRatioInformation const):
(WebCore::RenderReplaced::computeReplacedLogicalWidth const):
(WebCore::RenderReplaced::computeReplacedLogicalHeight const):
* Source/WebCore/rendering/RenderReplaced.h:
* Source/WebCore/rendering/RenderVideo.cpp:
(WebCore::RenderVideo::calculateIntrinsicSizeInternal):
(WebCore::RenderVideo::calculateIntrinsicSize):
(WebCore::RenderVideo::hasPosterFrameSize const):
* Source/WebCore/rendering/RenderVideo.h:

Canonical link: <a href="https://commits.webkit.org/261205@main">https://commits.webkit.org/261205@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/363c0be5d5c302d12707debfe11bfd175776febc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110696 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19778 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43258 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2050 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119529 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114663 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21186 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10890 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/1640 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116438 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15781 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98953 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103011 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97742 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30620 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44121 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12408 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31958 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/86008 "Found 1 new API test failure: /WebKitGTK/TestWebExtensions:/webkit/WebKitWebExtension/page-id (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12960 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8926 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18334 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51576 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7775 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14897 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Loirooriol commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -582,8 +580,8 @@ LayoutUnit RenderReplaced::computeReplacedLogicalWidth(ShouldComputePreferred sh
         bool computedHeightIsAuto = style().logicalHeight().isAuto();
-        bool hasIntrinsicWidth = constrainedSize.width() > 0;
-        bool hasIntrinsicHeight = constrainedSize.height() > 0;
+        bool hasIntrinsicWidth = constrainedSize.width() > 0 || shouldApplySizeOrInlineSizeContainment();
+        bool hasIntrinsicHeight = constrainedSize.height() > 0 || shouldApplySizeContainment();
```
I'm not familiar with this code, and I don't really see the logic behind treating size containment as having a positive intrinsic size. What would be the problem without this?

cathiechen replied:
The problem is that if hasIntrinsicWidth is false, the logical width might be calculated by other factors, like ratio * intrinsic height, etc. IIUC, that's not what we expected, for size containment boxes, the intrinsic size is specified, even it is 0, so it should be treated as if there is intrinsic size.

cathiechen replied:
It would affect some cases, like in contain-intrinsic-size-logical-003.html, 
```
  <canvas class="test cis-none" width="40" height="20"
       data-expected-client-width="0" data-expected-client-height="0"></canvas>
```

sammygill replied:
I think it would be nice to keep track of the intrinsic sizes and whether the dimensions are intrinsic together in their own structure. Then maybe we can modify it whenever we have things (like containment here) influence it and force one of the dimensions to become "intrinsic"

cathiechen replied:
Yeah, > 0 seems not accurate enough. Maybe we could use `std::optional` to store intrinsic size in the follow-up patch...

alanbaradlay reviewed (approved):
rs=me

cathiechen committed:
[contain-intrinsic-size] failures related to replaced elements
https://bugs.webkit.org/show_bug.cgi?id=253013

Reviewed by Alan Baradlay.

This patch makes intrinsicSize() considering both contain: size and inline-size,
so that there is no need to handle contain: inline-size specially.
In computeReplacedLogicalWidth/Height, hasIntrinsicWidth/Height is determined by width/height > 0,
for contain: size/inline-size, hasIntrinsicWidth/Height should be true regardless of the value.
For RenderVideo, in calculateIntrinsicSize(), the intrinsic size should consider contain: inline-size,
so this patch adjustSizeForInlineSizeContainment before return.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-028-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-logical-003-expected.txt:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::intrinsicLogicalWidth const):
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::computeAspectRatioInformationForRenderBox const):
(WebCore::RenderReplaced::computeIntrinsicRatioInformation const):
(WebCore::RenderReplaced::computeReplacedLogicalWidth const):
(WebCore::RenderReplaced::computeReplacedLogicalHeight const):
* Source/WebCore/rendering/RenderReplaced.h:
* Source/WebCore/rendering/RenderVideo.cpp:
(WebCore::RenderVideo::calculateIntrinsicSizeInternal):
(WebCore::RenderVideo::calculateIntrinsicSize):
(WebCore::RenderVideo::hasPosterFrameSize const):
* Source/WebCore/rendering/RenderVideo.h:

Canonical link: https://commits.webkit.org/261205@main
